standards/SPIFFE_Workload_API.md

```diff
@@ -0,0 +1,170 @@
+
+### 5.3 Default Identity
+
+It is often the case that a workload doesn’t know what identity it should assume. Determining when to assume what identity is a site-specific concern, and as a result, the SPIFFE specifications don’t reason about how to do this.
+
+In order to support the widest variety of use cases, the X.509-SVID Profile supports the issuance of multiple identities, while also defining a default identity. It is expected that workloads which are aware of multiple identities can handle decision making on their own. Workloads which don’t understand how to leverage multiple identities may use the default identity. The default identity is the first in the list. Protocol buffers ensure that the order of the list is preserved.
+
+### 5.4 Profile Messages
+
+The X.509-SVID Profile messages are expressed as a Protocol Buffer version 3 (proto3). They are defined below. For the specific service definition for this profile, please see the SPIFFE Workload API [Service Definition](#3-service-definition).
+
+```protobuf
+// The X509SVIDResponse message carries a set of X.509 SVIDs and their
+// associated information. It also carries a set of global CRLs, and a
+// TTL to inform the workload when it should check back next.
+message X509SVIDResponse {
+    // A list of X509SVID messages, each of which includes a single
+    // SPIFFE Verifiable Identity Document, along with its private key
+    // and bundle.
+    repeated X509SVID svids = 1;
+
+    // ASN.1 DER encoded
+    repeated bytes crl = 2;
+
+    // CA certificate bundles belonging to foreign Trust Domains that the
+    // workload should trust, keyed by the SPIFFE ID of the foreign
+    // domain. Bundles are ASN.1 DER encoded.
+    map<string, bytes> federated_bundles = 3;
+}
+
+// The X509SVID message carries a single SVID and all associated
+// information, including CA bundles.
+message X509SVID {
+    // The SPIFFE ID of the SVID in this entry
+    string spiffe_id = 1;
+
+    // ASN.1 DER encoded certificate chain. MAY include intermediates,
+    // the leaf certificate (or SVID itself) MUST come first.
+    bytes x509_svid = 2;
+
+    // ASN.1 DER encoded PKCS#8 private key. MUST be unencrypted.
+    bytes x509_svid_key = 3;
+
+    // CA certificates belonging to the Trust Domain
+    // ASN.1 DER encoded
+    bytes bundle = 4;
+
+}
+```
+
+
+All fields in the `X509SVID` message are mandatory, and MUST contain a non-default value. Clients receiving an `X509SVID` message in which any field has a default value SHOULD report an error and discard the message.
```

Comment:
Do we want to allow use of a Handshaker API or HSM/TPM to protect private keys? If so, can we make `x509_svid_key` optional?

Reply:
My thinking thus far has been that Handshaker API would be exposed over a dedicated gRPC service, separated from the X509SVID service. I'm under the impression that a Handshaker API would be able to accomplish TLS negotiation without necessarily tying itself to X509

If we protect a key pair with HSM/TPM, do you think we'd still use this X509SVID endpoint? Or perhaps we'd introduce a new service like Handshaker or something PKCS 11-shaped? I'm interested to hear more about this...

Reply:
Seems like we'd need to somehow indicate that the private key is intentionally mentioning and indicate that the Handshaker API is required. I think it's fine to tie to X509 here and (when support for non-exported key is needed) to add additional fields (ie. handshaker endpoint field or field for HSM key identity).

Reply:
Given the current behavior, if an administrator has earmarked an identity for non-exported keys only, a workload hitting X509SVID endpoint would receive a gRPC permission denied error. The workload would have to "know" that it needs to speak to a different endpoint (e.g. handshaker api). I like the idea of having a field which indicates that the key is missing because it is not exportable, however I don't think that this will work if we are also using gRPC error code.

If it's OK with you, I think we should address this when drafting the handshaker spec. I think we can probably update this text in a way that retains compatibility with existing non-handshake-aware clients.